### ISSUE TITLE
"Unsupported HW" Msg every time I start the PC

### ISSUE THREAD

**Issue by Shinji200** [zeffy/wufuc#158]
- wufuc version: v1.0.1.201
- OS: Windows 8.1 X64

Hi Zeffy, 
I want to report something, however I'm not entirely sure it's an issue, yesterday I installed the new version 1.0.1.201 but after that now every time I restart my PC I get the message of unsupported HW at the beginning of the session, I saw in the release notes that in rare occasions a similar behavior could be triggered but I'm not sure if it's normal to get it every time the windows session starts.

So far it seems I can still download updates and I don't see any errors in the log, also I haven't had the unsupported message after hours of using the PC, it's only at the start but can you tell me if this is normal? or should I wait a couple of days for the message to stop appearing?

Regards!

[wufuc.1.log](https://github.com/zeffy/wufuc/files/2020652/wufuc.1.log)

**Comment by zeffy** [comment]
I can't say I encountered it that often while testing, but yes it's probably normal. It likely depends on how the Windows Update service is configured to start up, etc. 

If you can open Windows Update and click "Check for updates" in the left pane of the window without getting another popup, then wufuc is working correctly. I know the occasional popup is annoying, but it was a trade off I had to make to have everything be more stable.

I'm looking into the possibility of adding something that waits for the Unsupported hardware popup to open and then just close it immediately, or prevent it from showing at all, but I still need to look into it further. Going to keep this issue open for now, until I get a better idea of if I'll be able to do that in a non-invasive way.

**Comment by Shinji200** [comment]
Hi Zeffy,
Thanks for the replay, for now it seems that I only get the message at the start of the session, also the windows update service was configured to start manually, just changed it to automatic delayed start, going to see if that changes things, seems I got unlucky as even after uninstalling wufuc and installing again I'm still getting the message hehe.

Regards

**Comment by ErnestoSF** [comment]
The same here, with wufuc version: v1.0.1.201 and Windows 8.1 X64, I receive the popup every time I start the computer or wakeup from sleep mode.

**Comment by zeffy** [comment]
So I can think of a couple ways to fix this, but neither seems perfect:

1. At the end of the patch procedure I can run some code that will essentially just close the popup if it is showing. However, the popup will usually flash on the screen for between a few milliseconds to a few seconds before it is closed, which could be jarring/confusing to the user if they don't know what is happening.
2. At user login, I can do the above to close any window that is already open, and then use `SetWindowsHookExW` to create a `WH_CBT` hook that will intercept **all** created windows, and then abort displaying the "Unsupported hardware" popup so that it will not show at all. The downside to this however, is that it is pretty intrusive. My `CBTProc` hook procedure will get injected into all applications that create a window in the user's session. Although in my experience it appears to be pretty stable, I don't know if it's worth it.

I'm still researching other ways to accomplish this, but these appear to have the most promise so far. Anyone care to share their thoughts?

**Comment by sammilucia** [comment]
@zeffy seeing as some people are using this in commercial environments I think the window hook method makes sense. Preventing it displaying you've either got to hook all windows or inject the WU process which from what you've said sounds like a moving target... So as far as dirty hacks go it's pretty clean lol. The only downside I can think of is it might trigger some antimalware software.

**Comment by Shinji200** [comment]
@zeffy so far I only get the pop-up at logon and not during "normal" use, it didn't change the behavior even after changing the startup type of the Windows Update service to automatic Delayed start. In any case I still can download updates without issues.

**Comment by zeffy** [comment]
@sammilucia Yeah it does work pretty well (at least in my test environment), but like you say I think some anti-malware (or possibly more concerning, anti-cheat systems in some games) would likely complain about it, which I think could end up being problematic. I'd really like to avoid system-wide hooks or injecting into system service processes unless there is no other way, so I'll keep doing research for now.

@Shinji200 If that is the case, then it seems like I could probably get away with doing a one-shot task that runs during user login and quickly closes the popup, should it exist. I will try doing this and see how well it works.

**Comment by sammilucia** [comment]
@zeffy yes, wufuc is essentially a hack to workaround a "feature" of Windows though, I can't think of another way.
What about just unhook as soon as it fires? So hook at startup, it suppresses the first warning then unhooks (maybe with a failsafe timeout). That could work while looking for other solutions.

**Comment by sammilucia** [comment]
@zeffy yes UnhookWindowsHookEx probably doesn't unload from DLLs which have already been hooked, just stops hooking future DLLs? I doubt this is a major problem though because it doesn't perform any function for those processes.
I suppose it depends on who your audience is for wufuc... If it's corporates who are trying to stay on Win 8.1 then ideally you wouldn't want the hardware warning to appear at all. If it's gamers/casual users then it's less of a problem (though it will probably create background noise of users who don't read the readme)... You could possibly provide a switch on install, but the hooking/unhooking may not even cause a problem in the wild.
What if you hook on startup, unhook after popup (and maybe after failsafe), and if no users report problems consider it closed?

**Comment by sammilucia** [comment]
+1 like it!

**Comment by rbknox** [comment]
Hi Zeffy, appreciate you clarifying that if updates are working without popup, then wufuc is working.  I had been running 1.0.0.191 and it was not clear if working because I was still getting frequent popups.   I have now installed 1.0.1.201 (Windows 8.1 x64 Kaby Lake) and am now NOT getting popups so far.  However, for both versions (191 and 201) I seem to get some installation errors (see attached - service has already been started and cannot find file specified errors).  Can you clarify whether these matter to correct operation of wufuc?   FYI, I am running Windows Media Center and popups appear on top of media center which is of course annoying while watching TV.  I thought I had WMC to be top layer but doesn't seem to work that way.   
![wufuc install screen](https://user-images.githubusercontent.com/39807431/40801588-1f94c6dc-64c8-11e8-9ff4-dbcadbcc7cf2.PNG)

**Comment by zeffy** [comment]
@rbknox those errors are normal, yes. During installation the script unconditionally tries to start the `Schedule` service and delete the wufuc scheduled task if it already exits. So if the service is already started or the wufuc task doesn't exist, then it will spit out those errors. I've been meaning to silence them but haven't had a chance. I'll try to get that integrated into the script by the next version, thanks for reminding me!

**Comment by sammilucia** [comment]
I've seen the Unsupported Hardware window several times waking from sleep. Not consistently though, it happens probably 1/4 to 1/3 the time. It may be related to a timing issue (the UH window is called just as the PC is waking up). I doubt it's related to different sleep states looking through the System Event Log. As usual there's no problems installing updates.

It might be a case to keep the hardware hook installed, if you decide to go that route?

**Comment by zeffy** [comment]
Could also probably register callbacks that set the hook/event proc with a timeout when connecting to a user session (like logging in or waking up the pc). I believe this should be possible.

For now though I'm looking into other options. I've found a couple potential routes I could go, such as locking one of WU's global named mutexes or events so that it can't proceed until I unlock it, or intercepting the WU client directive that triggers the popup. Haven't figured exactly out how to do that yet, but it may be possible.

Sorry for the slow updates on this btw, I've been dealing with computer hardware issues that have been preventing me from doing much coding the past few days.

**Comment by zeffy** [comment]
Also worth noting, I believe I've figured out how to solve this without any ugly hacks on Windows 8.1. It has a newer function available that I can use in place of `NotifyServiceStatusChange` called [`SubscribeServiceChangeNotifications`](https://msdn.microsoft.com/en-us/library/windows/desktop/hh973487(v=vs.85).aspx), which appears to be ran synchronously. I haven't tested it yet, but if it works like I think it does then there would be no race condition that would allow for the popup to be displayed in the first place.

That still leaves Windows 7, though.

**Comment by sammilucia** [comment]
That sounds good for w8.1 (which I use)... I still don't think the original hook was that dirty 😂. Sorry about the PC hardware problems 😬

**Comment by Kenmercermusic** [comment]
I installed a version of Wufuc last year and everything was working for me until a couple of days ago, I then tried to install the latest version but still getting the 'Unsupported Hardware' pop-up. It even appears when I go to windows update and select 'Check For Updates'. Any advice/fix for this would be greatly appreciated.

**Comment by zeffy** [comment]
@Kenmercermusic could you upload your log from `C:\ProgramData\wufuc\wufuc.1.log`? thanks

**Comment by shak3800** [comment]
Hi there

I seem to be getting as well the popup very often using windows 8.1 . Is there a way to disable it or a workaround?

**Comment by Drayke-Halcyon** [comment]
First of all, I'd like to say thank you for devoting your time and energy to this project. I'm a huge fan!

That being said, I've had a system running Windows 7 Professional (x64), and haven't had any troubles since it was installed, until the last couple of days. I'm getting the pop-up when booting the system now, as well as when we run Windows Update. Is there something I'm overlooking?

**Comment by arrowgamer** [comment]
I'm getting the popup after booting. I'm using a z370 motherboard.

**Comment by shak3800** [comment]
This is probably a long shot. Is it possible you couls spoof the cpu/gpu to display supported models at least when on boot in order to pass the check?

**Comment by zeffy** [comment]
@shak3800 Unfortunately no. The CPU check uses the `cpuid` assembly instruction which queries the processor directly about its specs. It is possible if you are running Windows inside of a hypervisor (which is between the OS and the CPU), like vmware or vbox, but for most people that is not an option. Also in regards to your previous question, you could try an older version but I don't recommend it.

 @Drayke-Halcyon you shouldn't be getting any popup when you check for updates, can you post your log file? Thanks.

@arrowgamer as long as you aren't getting a popup after that, that's normal. I'm working on this atm.

**Comment by Kenmercermusic** [comment]
@zeffy Hi Zeffy, I uploaded my log file 2 weeks ago as you requested but you never got back to me? Just wondering if there is some solution to my issue, it's really getting on my nerves now as it pops up frequently when I'm trying to work. Any help would be appreciated. Thanks. Ken

**Comment by zeffy** [comment]
@Kenmercermusic sorry, I had been having pc hardware issues as noted above, and have had a lot of other stuff going on in my life as well, so I haven't been able to spend as much time on this project. I took a look at the log file you posted and there was nothing out of the ordinary other than it appearing to have only ran one time on 6/8/18. I figure whatever is going on will likely be resolved once I fix the main issue this thread is about, there isn't much that can be done about it without an update, which I am actively working on.

**Comment by Kenmercermusic** [comment]
@zeffy It seems to be getting worse for me, at first it was just one time on restart but now it randomly pops up several times throughout the day. I really appreciate your commitment to helping us all with the new upgrade you are working on. Thank you!

**Comment by donmega** [comment]
I am also keep getting the unsupported hw message. I've attached the log file (Windows 8.1 x64)

[wufuc.1.log](https://github.com/zeffy/wufuc/files/2132088/wufuc.1.log)

**Comment by UnbreakCode** [comment]
@Kenmercermusic I think the the Msg comes,when Windows is trying to automatically search for Updates.You can disable the automatic Function for searching Windows Updates as a little fix for the random popup.

**Comment by Kenmercermusic** [comment]
@zeffy Thanks, I'll give it a try for a few days and let you know how it goes!

**Comment by remco8264** [comment]
Setting the WU Service to **disabled** makes the popup dissappear at boot.
If you want to install update, you'll need to enable the service, install the updates, and disable the service again afterwards.

**Comment by GCRaistlin** [comment]
Same problem here.
[wufuc.1.log](https://github.com/zeffy/wufuc/files/2329760/wufuc.1.log)

**Comment by GCRaistlin** [comment]
**@zeffy**, I found out that it's because of DrWeb. The problem is that I couldn't find the workaround - even if all protection components are disabled the error is the same. BTW is the log file being created only by the installation? If I delete it it doesn't appear again after the reboot.

**Comment by zeffy** [comment]
>BTW is the log file being created only by the installation? If I delete it it doesn't appear again after the reboot.

The log file should be written to whenever wufuc starts, and when the Windows Update service starts, if it isn't being created then odds are the task isn't starting at all (or something is interfering with it working). 

In regards to DrWeb causing the issue, I'm afraid that I don't have the time/resources to provide support for security configurations besides Windows Defender or MS Security Essentials. With anything else, you're pretty much on your own unless another user can explain how to make it work along side wufuc.
